### PR TITLE
PCiOS: Update Fastfile with version model

### PR DESCRIPTION
**Pull request by spencertransier**
## Summary
- This PR integrates the new version and build code models from https://github.com/wordpress-mobile/release-toolkit/pull/512
- Many of the changes here are from breaking out the code within the various Release Toolkit `*_prechecks` actions into the Fastfile itself. Nearly all of the `prechecks` actions were tightly coupled with the existing version methods. To simplify things with the new version model, I took the relevant snippets from the `prechecks` actions and added them directly to the Fastfile. We'll be evaluating those and adding better-named and succinct shared actions to the Release Toolkit in the future. 
- "Release version" refers to the `VERSION_SHORT` in the xcconfig file: either the `1.2` or `1.2.3` (hotfix) version
-  "Build code" is what we're calling the incremented number for each build (in PCiOS's case, the format of `1.2.3.4`, `VERISON_LONG` in the xcconfig file.). This is because the `versionCode` in Android apps or the `build_number` used by other A8C apps are formatted differently. So "build code" is the overarching term for all of those. Just wanted to clarify that vocabulary.

### CODE REVIEW

File: fastlane/Fastfile
Comment (jkmassel) [inline fastlane/Fastfile]: Can we put this at the top with the other check?

**Review by jkmassel — DISMISSED** [review]

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: Addressed in https://github.com/Automattic/pocket-casts-ios/pull/1137/commits/b6826af6e16656e8c8e513b10e5d66c6147b3c3a

File: fastlane/Fastfile
Comment (pocketcasts) [inline fastlane/Fastfile]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/c13a215366b1c2abb0171201a62dc6a39267e30b/fastlane/Fastfile#L249"></span>Useless assignment to variable - <code>confirmation_message</code>.</td>
    </tr>
  </tbody>
</table>

File: fastlane/Fastfile
Comment (pocketcasts) [inline fastlane/Fastfile]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/c13a215366b1c2abb0171201a62dc6a39267e30b/fastlane/Fastfile#L559"></span>Inconsistent indentation detected.</td>
    </tr>
  </tbody>
</table>

File: fastlane/Fastfile
Comment (pocketcasts) [inline fastlane/Fastfile]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/c13a215366b1c2abb0171201a62dc6a39267e30b/fastlane/Fastfile#L560"></span>Inconsistent indentation detected.</td>
    </tr>
  </tbody>
</table>

Comment (spencertransier) [reply fastlane/Fastfile]: Fixed in https://github.com/Automattic/pocket-casts-ios/pull/1137/commits/d951f537ecef9223930535e1d1bc9bf9a14b97f8

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: Fixed in https://github.com/Automattic/pocket-casts-ios/pull/1137/commits/d951f537ecef9223930535e1d1bc9bf9a14b97f8

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: Fixed in https://github.com/Automattic/pocket-casts-ios/pull/1137/commits/d951f537ecef9223930535e1d1bc9bf9a14b97f8

**Review by mokagio — COMMENTED** [review]
Just a note that I've been focusing on https://github.com/wordpress-mobile/WordPress-iOS/pull/21666 and leaving various suggestions there. 

I think it'd be best to lock in the implementation on that end, then propagate the learnings here.

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile]:
```suggestion
    UI.user_error!('Aborted by user request') unless options[:skip_confirm] || UI.confirm('Do you want to continue?')
```

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile]: Super nitpick
```suggestion
    UI.success "Done! New Release Version: #{release_version_current}. New Build Code: #{build_code_current}."
```

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile]:
```suggestion
    UI.user_error!('Aborted by user request') unless options[:skip_confirm] || UI.confirm('Do you want to continue?')
```

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile]:
```suggestion
    UI.user_error!('Aborted by user request') unless options[:skip_confirm] || UI.confirm('Do you want to continue?')
```

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile]:
```suggestion
    UI.user_error!('Aborted by user request') unless options[:skip_confirm] || UI.confirm('Do you want to continue?')
```

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile]: This has been removed from the WordPress iOS version, https://github.com/wordpress-mobile/WordPress-iOS/pull/21666 , on the basis that beta release are made on top of the current release branch.

Thinking about it, I think there was some logic / expectation that new betas should start only if the previous beta has been merged. I say this because I run in that scenario a number of times in WordPress and had to restart the lane passing the build code I wanted to use, which resulted in it ignoring what was on `trunk` and using `release/x.y` as the source of truth. I think the process we devised in WordPress iOS is leaner and would argue for using it here, too.

Maybe worth seeing how shipping betas for WordPress 23.6 goes?

Comment (spencertransier) [reply fastlane/Fastfile]: Yeah, I'd say let's wait and see if there are any issues with WP before we rollout that change to the other apps.

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile]: [In WordPress-iOS](https://github.com/wordpress-mobile/WordPress-iOS/blob/dfa631f130ee22498aaf920ddb49ace9dbe6811b/fastlane/lanes/release.rb#L467-L473) I too started with `git_add` and `git_commit` but then [removed it](https://github.com/wordpress-mobile/WordPress-iOS/commit/69a7414f2148c6ca96e5f82d06c4bd2b2fb565b1) because Fastlane's `git_commit` also adds the files.

The reason I thought to use `git_add` was because my `git_commit` call failed the first time. But that turned out to be due to case sensitivity in the files path.

**Review by mokagio — APPROVED** [review]
Approving to unblock. 

In the comments, we agree to wait and see how betas behave in WordPress, but I haven't had a chance to make one yet 😅

File: fastlane/Fastfile
Comment (leandroalonso) [reply fastlane/Fastfile]: Shall we remove `git_add` then?

**Comment by leandroalonso** [comment]
@spencertransier JFYI I fixed a few merge conflicts. Those were simple so I think it would be worth it before I review the new code.

File: fastlane/Fastfile
Comment (leandroalonso) [inline fastlane/Fastfile]: I know this already happens, but why do we need to check and pull `trunk` when releasing a new beta?

**Review by leandroalonso — COMMENTED** [review]
@spencertransier it all looks fine to me!

Left a few comments though just to make sure I'm following all the changes before we merge it :)

File: fastlane/Fastfile
Comment (mokagio) [reply fastlane/Fastfile]: Ah, good question. Thank you @leandroalonso 

From top of memory, the previous implementation of the version bump would look for the value on `trunk` to determine which new beta version to make. Therefore, we needed beta 1 merged into `trunk` to start beta 2, or we'd have to specify the new beta version manually.

@spencertransier might be able to give more details on whether the new process still requires that. If not, we might as well avoid the dance.

File: fastlane/Fastfile
Comment (mokagio) [reply fastlane/Fastfile]: Just a note to say that I finally deployed a beta for Jetpack and WordPress iOS and it all worked fine 😄

File: fastlane/Fastfile
Comment (leandroalonso) [reply fastlane/Fastfile]: I think it's not a bad thing to have, just wanted to know. :)

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: Good to know! Removed in https://github.com/Automattic/pocket-casts-ios/commit/68bcf786a4a954a4325dff56064236faf1ed7a11

File: fastlane/Fastfile
Comment (spencertransier) [reply fastlane/Fastfile]: @leandroalonso @mokagio The new tooling does not need to check `trunk` out first, so I replaced this line with a `release/` branch check instead: https://github.com/Automattic/pocket-casts-ios/commit/89f08f295c64e6cc5bdfecb49a053f30356ff07b

File: fastlane/Fastfile
Comment (AliSoftware) [reply fastlane/Fastfile]: Calling `git_add` before `git_commit` is only necessary for brand new files that were not under version control until now. If the file was already tracked in git and all you wanted to do is stage the changes of the existing file, just calling `git_commit` is enough.

iinm, this behavior is consistent with `git add` and `git commit -a` (or `git commit {list of files}`) in `git` itself, not really specific to those _fastlane_ actions. For example `git commit -a` will add all the already-tracked-but-modified files to the commit you're creating, but won't add any new file that wasn't tracked in git yet, and you'll have to call `git add` explicitly on those new files first if you want to start tracking them and include them in your commit.

File: fastlane/Fastfile
Comment (leandroalonso) [reply fastlane/Fastfile]: @spencertransier I just ran into an "issue" and I guess we need it? What just happened:

1. Released a new beta
2. Went to release another one before the current changes were merged into `trunk`
3. It prompted me if I'd like to release a new beta, however the build number was not incremented

I was puzzled as to why and it was because the commit that the tag refers to needs to be on `trunk` apparently (in my case the PR wasn't merged, but I wonder if we don't make sure `trunk` is updated this issue will happen).

Comment (spencertransier) [reply fastlane/Fastfile]: @leandroalonso Were the beta builds both of the same version (like 7.52.0.1 then 7.52.0.2)? Or on different release branches?

And the release branches are branching off of this PR's branch with the new versioning tooling?

**Comment by pocketcasts** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/Automattic/pocket-casts-ios/milestone/33">7.50</a>. The due date for this milestone has already passed.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: fastlane/Fastfile
Comment (leandroalonso) [reply fastlane/Fastfile]: @spencertransier This is going to be a bit too much but I think it gives enough detail to try to understand what happened:

1. I released the beta `7.52.0.1`
2. I then realized #1240 was merged into `trunk`
3. We then opened #1243 
4. In the meantime, `7.52.0.1` was ready and #1245 was open and was waiting for CI build to pass
5. I then started the new beta, but it said the new beta would be `7.52.0.1` (it should have been `7.52.0.2`)
6. I waited till #1245 was merged into `trunk`
7. Tried again and this time it said the new beta would be `7.52.0.2`, like it should

You can reproduce this right now if you want:

1. #1247 is not merged
2. Go to `release/7.53` and then `bundle exec fastlane new_beta_release`
3. What is expected: the next beta to be `7.52.0.3`, instead it will say it will be `7.52.0.2`

That's why I think we need to update `trunk`, otherwise it might affect the beta version?

Comment (spencertransier) [reply fastlane/Fastfile]: @leandroalonso Thanks for the details! 

That behavior is expected because this PR is not merged yet. The `new_beta_release` lane in the `release/7.53` branch still calls the `ios_betabuild_prechecks` Fastlane action, which does switch back to `trunk` and uses the `trunk` version number to determine the next beta version number. After this PR is merged, the Fastfile won't use the `ios_betabuild_prechecks` action anymore and won't need to go to trunk first before creating a new beta release. With this PR, the tooling checks what the version is in the `release/x.x` branch to determine the next beta build number.

Comment (leandroalonso) [reply fastlane/Fastfile]: Ohhhhhh got it!

Thanks for the explanation @spencertransier!

Comment (spencertransier) [reply fastlane/Fastfile]: @leandroalonso You're welcome! 😃 Are you good with me merging this PR now?

**Review by leandroalonso — APPROVED** [review]

**Pull request merged**
